codec/src/main/java/io/netty/handler/codec/compression/Bzip2Compressor.java

```diff
@@ -114,48 +114,52 @@ private Bzip2Compressor(final int blockSizeMultiplier) {
 
     @Override
     public ByteBuf compress(ByteBuf in, ByteBufAllocator allocator) throws CompressionException {
-        if (compressorState != CompressorState.PROCESSING) {
-            return Unpooled.EMPTY_BUFFER;
-        }
-
-        ByteBuf out = allocator.buffer();
-
-        for (;;) {
-            switch (currentState) {
-                case INIT:
-                    out.ensureWritable(4);
-                    out.writeMedium(MAGIC_NUMBER);
-                    out.writeByte('0' + streamBlockSize / BASE_BLOCK_SIZE);
-                    currentState = State.INIT_BLOCK;
-                    // fall through
-                case INIT_BLOCK:
-                    blockCompressor = new Bzip2BlockCompressor(writer, streamBlockSize);
-                    currentState = State.WRITE_DATA;
-                    // fall through
-                case WRITE_DATA:
-                    if (!in.isReadable()) {
-                        return out;
-                    }
-                    Bzip2BlockCompressor blockCompressor = this.blockCompressor;
-                    final int length = Math.min(in.readableBytes(), blockCompressor.availableSize());
-                    final int bytesWritten = blockCompressor.write(in, in.readerIndex(), length);
-                    in.skipBytes(bytesWritten);
-                    if (!blockCompressor.isFull()) {
-                        if (in.isReadable()) {
+        switch (compressorState) {
+            case CLOSED:
+                throw new CompressionException("Compressor closed");
+            case FINISHED:
+                return Unpooled.EMPTY_BUFFER;
+            case PROCESSING:
+                ByteBuf out = allocator.buffer();
+                for (;;) {
+                    switch (currentState) {
```

Comment:
How about extracting the contents of this `case` to its own method? It would reduce indentation and make the code easier to follow, I think.